quickstart/daml/licensing-tests/daml/Licensing/Scripts/TestLicense.daml

```diff
@@ -0,0 +1,217 @@
+      installMeta = Metadata with
+        values = Map.fromList [("providerUserId", "<user-id>")]
+
+-- | Simulate the setup of the validator party on a self-hosted validator node.
+setupValidatorParty : AmuletApp -> Text -> Script AmuletUser
+setupValidatorParty app name = do
+  party  <- allocatePartyWithHint name $ PartyIdHint name
+  userId <- validateUserId name
+  -- We need 'CanReadAs app.dso' only in tests, as they are not yet setup to simulate explicit disclosure.
+  -- In the actual implementation, the provider backend would provide the DSO published contracts via explicit disclosure.
+  createUser (User userId (Some party)) [CanActAs party, CanReadAs app.dso]
+  -- the validator party is its own validator and can thus claim all its validator rewards
+  recordValidatorOf app party party
+  pure $ AmuletUser userId party
+
+
+setupLicense : LicensingApp -> AmuletUser -> Script (ContractId License)
+setupLicense app user = do
+    installId <- getInstall
+    -- need: readAs dso for explicit disclosure of AmuletRules and OpenMiningRound
+    result <- submitMulti [app.provider.primaryParty] [app.dso] do
+      exerciseCmd installId AppInstall_CreateLicense with
+        params = licenseParams
+    pure result.licenseId
+  where
+    getInstall : Script (ContractId AppInstall)
+    getInstall = do
+      [(installId, _)] <- query @AppInstall user.primaryParty
+      pure installId
+    licenseParams = LicenseParams with
+      meta = Metadata with
+        values = Map.fromList [("licenseId", "<dummy-uuid>")]
+
+renewLicense : LicensingApp -> ContractId License -> Decimal -> RelTime -> Script (ContractId LicenseRenewalRequest, ContractId AppPaymentRequest)
+renewLicense app licenseId licenseFeeCc licenseExtensionDuration = submit app.provider.primaryParty do
+  exerciseCmd licenseId License_Renew with
+    licenseFeeCc
+    licenseExtensionDuration
+    paymentAcceptanceDuration = days 1
+    description = "Renewal of the license"
+
+
+completeTransfer : AmuletApp -> AmuletUser -> ContractId License -> ContractId AcceptedAppPayment -> Script ()
+completeTransfer app provider licenseCid acceptedPaymentCid = do
+  transferContext                     <- getAppTransferContext app provider
+  Some AcceptedAppPayment{reference}  <- queryContractId provider.primaryParty acceptedPaymentCid
+  [(renewalRequestCid, _)]            <- 
+    queryFilter @LicenseRenewalRequest provider.primaryParty $ \c -> c.reference == reference
+  void $ submitUser provider.userId do 
+    exerciseCmd renewalRequestCid LicenseRenewalRequest_CompleteRenewal with ..
+
+rejectPaymentRequest : LicensingApp -> AmuletUser -> ContractId LicenseRenewalRequest -> ContractId AppPaymentRequest -> Script ()
+rejectPaymentRequest app sender renewalRequest paymentRequest = do
+  submit sender.primaryParty do exerciseCmd paymentRequest AppPaymentRequest_Reject
+  void $ submitMulti [app.provider.primaryParty, sender.primaryParty] [] do archiveCmd renewalRequest
```

Comment:
```suggestion
  void $ submit sender.primaryParty $ do archiveCmd renewalRequest
```
Important: multi party submits are a code smell. Here were doing something that wouldn't work in production: the provider and user would never submit a command jointly. It is one or the other acting.